apps/server/src/modules/oauth-provider/service/oauth-provider.login-flow.service.ts

```diff
@@ -1,48 +1,34 @@
-import { ProviderLoginResponse } from '@shared/infra/oauth-provider/dto';
-import { LtiToolDO } from '@shared/domain/domainobject/ltitool.do';
-import { LtiToolRepo, PseudonymsRepo } from '@shared/repo';
-import { InternalServerErrorException } from '@nestjs/common';
-import { Permission, PseudonymDO, User } from '@shared/domain';
 import { Injectable } from '@nestjs/common/decorators/core/injectable.decorator';
-import { AuthorizationService } from '@src/modules/authorization';
+import { NotFoundException } from '@nestjs/common/exceptions/not-found.exception';
+import { ExternalToolDO } from '@shared/domain';
+import { LtiToolDO } from '@shared/domain/domainobject/ltitool.do';
+import { LtiToolService } from '@src/modules/lti-tool/service';
+import { ExternalToolService } from '@src/modules/tool/service';
 
 @Injectable()
 export class OauthProviderLoginFlowService {
 	constructor(
-		private readonly ltiToolRepo: LtiToolRepo,
-		private readonly pseudonymsRepo: PseudonymsRepo,
-		private readonly authorizationService: AuthorizationService
+		private readonly ltiToolService: LtiToolService,
+		private readonly externalToolService: ExternalToolService
 	) {}
 
-	async getPseudonym(currentUserId: string, loginResponse: ProviderLoginResponse): Promise<PseudonymDO> {
-		if (!loginResponse.client.client_id) {
-			throw new InternalServerErrorException('Could not find oAuthClientId in login response to get the pseudonym');
+	public async findToolByClientId(clientId: string): Promise<ExternalToolDO | LtiToolDO> {
+		const externalTool: ExternalToolDO | null = await this.externalToolService.findExternalToolByOAuth2ConfigClientId(
+			clientId
+		);
+		const ltiTool: LtiToolDO | null = await this.ltiToolService.findByClientIdAndIsLocal(clientId, true);
+
+		if (!externalTool && !ltiTool) {
+			throw new NotFoundException(`Unable to find ExternalTool or LtiTool for clientId: ${clientId}`);
 		}
 
-		const ltiToolDO: LtiToolDO = await this.ltiToolRepo.findByClientIdAndIsLocal(loginResponse.client.client_id, true);
-		const pseudonym: PseudonymDO = await this.pseudonymsRepo.findByUserIdAndToolId(
-			currentUserId,
-			ltiToolDO.id as string
-		);
-		return pseudonym;
+		return (externalTool ?? ltiTool) as ExternalToolDO | LtiToolDO;
```

Comment:
Can you please remove this cast, or replace it with a tryToCastTo... private method? You can also use a typescript typeguard it work well and we have some examples already in code. Maybe with this the as cast can be avoided.

function isNumber(x: any): x is number {
  return typeof x === "number";
}

We have examples in code, but it is hard to search "is" isnt the best key word to find it. :D